### ISSUE TITLE
Salt-cloud unable to bootstrap minions using COPR repo

### ISSUE THREAD

**Issue by saint0192** [saltstack/salt#25886]
This just started this morning using salt-cloud to bootstrap AWS minions:

I had no issues last night while working with the same systems and commands

Symptoms: Salt-cloud unable to finish bootstrap of any minions

How to reproduce: salt-cloud -p <profile> <tag>

Master: Ubuntu 14.04 server

Minion: current Amazon Linux AMI

Log from minion:

# tail /var/log/salt/minion 
2015-07-30 18:15:08,091 [salt.cli.daemons ][INFO    ][2801] Setting up the Salt Minion "7-30-update-2"
2015-07-30 18:15:08,183 [salt.cli.daemons                         ][INFO    ][2801] The salt minion is shut down
2015-07-30 18:15:08,185 [salt.cli.daemons                         ][INFO    ][2807] The salt minion is shut down
2015-07-30 18:15:08,323 [salt.minion                              ][ERROR   ][2809] Invalid keyword 'str' for variable 'master_type'
2015-07-30 18:15:08,323 [salt.cli.daemons                         ][INFO    ][2809] The salt minion is shut down

Repository used by minion to install salt-minion:

# cat /etc/yum.repos.d/saltstack-salt-epel-6.repo 
[saltstack-salt]
name=Copr repo for salt owned by saltstack
baseurl=https://copr-be.cloud.fedoraproject.org/results/saltstack/salt/epel-6-$basearch/
skip_if_unavailable=True
gpgcheck=1
gpgkey=https://copr-be.cloud.fedoraproject.org/results/saltstack/salt/pubkey.gpg
enabled=1

I have not installed any developer branch software to test for this bug. This issue was introduced into the repo within the last day or two as my systems have remained updated regularly and this problem did not exist yesterday...

Let me know if any other info needed. Thanks.

**Comment by mooperd** [comment]
Hitting exactly the same problem with salt-minion 2015.5.3 (Lithium)

**Comment by saint0192** [comment]
My version is 2015.5.2

**Comment by mooperd** [comment]
Nothing very exciting in the debug log.
```
[root@ip-10-101-10-36 centos]# salt-minion -l trace
[DEBUG   ] Reading configuration from /etc/salt/minion
[TRACE   ] None of the required configuration sections, 'logstash_udp_handler' and 'logstash_zmq_handler', were found the in the configuration. Not loading the Logstash logging handlers module.
[DEBUG   ] Configuration file path: /etc/salt/minion
[TRACE   ] Trying pysss.getgrouplist for 'root'
[TRACE   ] Trying generic group list for 'root'
[TRACE   ] Group list for user 'root': []
[INFO    ] Setting up the Salt Minion "build061"
[DEBUG   ] Created pidfile: /var/run/salt-minion.pid
[DEBUG   ] Reading configuration from /etc/salt/minion
[TRACE   ] Device xvda reports itself as an SSD
[ERROR   ] Invalid keyword 'str' for variable 'master_type'
[INFO    ] The salt minion is shut down
```

**Comment by saint0192** [comment]
It is the invalid master type that is killing it. I did see another bug/issue files a few months ago that appears to be the same...

https://github.com/saltstack/salt/issues/25335

Is this a re-appearance of the same issue? Or a delayed onset?

**Comment by mooperd** [comment]
Removing "master_type: str" from the config file does indeed allow the salt-minion to load.

**Comment by saint0192** [comment]
Yes. I just did the same, and it allows the minion to finish loading. I hope they can put in a fix soon :-( because I'm unable to make automated builds right now. This is the worst possible time for this to break... At least for me. I'm in a crunch as my company's product is getting ready for Beta. I need the automation of our AWS cloud to be finalized soon so it can be tested during Beta and fairly polished within a few months for the actual release to consumers...

This stops me completely...

**Comment by mooperd** [comment]
@saint0192 

In your cloud.profiles you can specify that a specific version of salt should be installed from git. This protects you from hitting these kinds of problems. Because its obviously quite hard to do: Salt-cloud, bootstrap-salt and saltstack do not seem to get adequate integration testing. 

/etc/salt/cloud.profiles
```
aws-amazon:
    provider: aws
    image: ami-1624987f
    size: t1.micro
    ssh_username: ec2-user
    script: bootstrap-salt
    script_args: git v2015.2
```

**Comment by saint0192** [comment]
Excellent. Thanks. I'll give that a whirl to see if I can back-rev so I can continue working.

Thanks!

**Comment by saint0192** [comment]
Well... This got worse. I can't use 2015.2 or 2015.3 because they are broken. Using your recommendation I back-revved to 2014.7.6 because that was the last version I could find before 2015.2 and the bootstrap-salt.sh as well as the /etc/rc.d/init.d/salt-minion it puts on the minion-to-be is broken as well... Its path for salt-minion is wrong... See below:

'''
Starting salt-minion daemon: /bin/bash: /usr/bin/salt-minion: No such file or directory
                                                           [FAILED]
 *  INFO: Running daemons_running()
 * ERROR: salt-minion was not found running
 * ERROR: Failed to run daemons_running()!!!
 * ERROR: salt-minion was not found running. Pass '-D' to bootstrap-salt.sh when bootstrapping for additional debugging information...
[root@ip-172-31-45-180 ec2-user]# find / -name salt-minion
/usr/local/bin/salt-minion
/etc/rc.d/init.d/salt-minion
'''

**Comment by mooperd** [comment]
I just installed a salt master on Centos7

```
#salt-bootstrap installs salt-master and salt-minion
curl -L https://bootstrap.saltstack.com -o install_salt.sh
sh install_salt.sh -M git v2015.8
```

I was then able to launch a minion with the following cloud.profile.

```
centos7_ec2_nginx:
    provider: ec2-eu-west-1
    image: ami-e4ff5c93
    size: t2.medium
    script: bootstrap-salt
    script_args: git v2015.8
    subnetid: subnet-XXXX
    securitygroupid:  sg-XXXX
```

This was all on the stock CentOS 7 x86_64 image available in the marketplace.

**Comment by saint0192** [comment]
I still have the same problem with that version. I am using the Amaazon Linux AMI and salt-bootstrap is installing the salt-minion script to the wrong location (/usr/local/bin instead of /usr/bin). Why is it doing this?

<pre><code>
Starting salt-minion daemon: /bin/bash: /usr/bin/salt-minion: No such file or directory
                                                           [FAILED]
 *  INFO: Running daemons_running()
 * ERROR: salt-minion was not found running
 * ERROR: Failed to run daemons_running()!!!
 * ERROR: salt-minion was not found running. Pass '-D' to bootstrap-salt.sh when bootstrapping for additional debugging information...
[root@ip-172-31-40-147 ec2-user]# updatedb
bash: updatedb: command not found
[root@ip-172-31-40-147 ec2-user]# find / -name salt-minion
/usr/local/bin/salt-minion
/etc/rc.d/init.d/salt-minion
</code></pre>

I was able to re-create a base Amazon Linux image for my minions with /etc/default/salt to include:

<pre><code>
SALTMASTER=/usr/local/bin/salt-master
SALTMINION=/usr/local/bin/salt-minion
</code></pre>

As shown in:

https://github.com/saltstack/salt/issues/25692

So I am back up and running. The issues should be fixed, but I'm comfortable with the solution I have for now and back in action.

Thanks for your help @mooperd

**Comment by rallytime** [comment]
Sorry for the trouble here @saint0192 and @mooperd as this is definitely a frustrating bug. This issue is a duplicate of #25335 and has already been fixed by @cachedout in #25703 and #25704. The fix will be available in the next release of salt, which will either be 2015.5.4 or 2015.8.0, whichever one comes out first.

**Comment by saint0192** [comment]
Thanks for the response @rallytime. I appreciate your product and your effort. If it is any consolation, we are running through furious bug-fix before the impending Beta of our product right now :-)

http://blinkforhome.com/

So I feel your pain...

**Comment by rallytime** [comment]
@saint0192 That looks cool - Good luck! It's definitely a crazy process. :)

Is it alright with you if we close this issue? Or is there anything that I missed in here that still needs resolution?

**Comment by saint0192** [comment]
You can close the issue if the fix is in place and just needs to propagate out.

The original issue was that a bootstrapped minion on Amazon Linux cannot recognize the master type "str"

the secondary issue was the wrong path for salt-minion on Amazon Linux. Of course the common factor is Amazon Linux, and if that is the base of the problem I may need to move to Centos at some point...

**Comment by rallytime** [comment]
@saint0192 Sounds great. Yes, the original issue with `master_type: str` has been fixed. It's not in a release yet, but it will be soon.

The second issue with Amazon Linux is specific to the bootstrap script itself, which actually has it's own repo. The bug you linked to above in the SaltStack repo was moved to the bootstrap repo here: https://github.com/saltstack/salt-bootstrap/issues/636 and has been fixed by https://github.com/saltstack/salt-bootstrap/pull/638. Note, however, that the fix is only on the `develop` version of the bootstrap script. It will be available in the next stable release of `salt-bootstrap`. 

You can access the develop version of the bootstrap script, if you'd like, [here](https://raw.githubusercontent.com/saltstack/salt-bootstrap/develop/bootstrap-salt.sh).

Thanks again for these reports and for the testing! I'll go ahead and close this, but do let us know if you run into more problems.

**Comment by marek-obuchowicz** [comment]
@rallytime is there any way that salt-cloud will be tested a little better before it is released? We really appreciate a lot of new features and updates in salt, but it's really frustrating that so often there are bugs, that render the whole salt-cloud unusable (like with AssociatePublicIpAddress, now problems with bootstrapping)... We come one day to the office, run "salt-cloud" as usual and find, that it doesn't work :(

**Comment by mooperd** [comment]
@marek-obuchowicz how are you installing salt on your master and minions? from git or epel?

**Comment by marek-obuchowicz** [comment]
I am using debian (so no epel) and have been using, till now, the default options of salt-cloud / bootstrap-salt. At the moment switched for git, workaround with using fixed tag 2015.5.2 works fine (but makes bootstrapping much slower)

**Comment by claudiupopescu** [comment]
@rallytime as we discussed this in the past, is there any progress in the direction of a hot release procedure?
Breaking vital parts of salt and leaving it like this for weeks is not ok :)

**Comment by rallytime** [comment]
@claudiupopescu We are still working very hard on ironing out our release schedules and processes. I believe we're releasing 2015.5.4 very soon.

**Comment by ymarson** [comment]
Is there an eta for this. We are using salt on AWS and cannot auto-scale as we need to manually remove the master_type field to start the salt-minions

**Comment by mooperd** [comment]
I filed a bug on the Redhat bugzilla for EPEL last week. It was assigned to @herlo.  https://bugzilla.redhat.com/show_bug.cgi?id=1249117

**Comment by marek-obuchowicz** [comment]
Please note that the problem is present not only on EPEL-based distributions - I had it on debian 8 when using bootstrap_salt without any parameters (so latest stable release, 2015.5.3).

**Comment by complistic-gaff** [comment]
Also had problems with 2015.5.3 on Rackspace Debian 8 instances. I got around it by adding this to my `/etc/salt/cloud.providers.d/rackspace.conf`.

```
rackspace:
  ...
  minion:
    master_type: standard
```

This changes the config on the bootstraped minion to use `standard` and not `str` and all is right with the world again.

Not sure what will happen once the fix lands in 2015.5.4 though... I might need to revert that config.

**Comment by dannykansas** [comment]
@complistic-gaff Your fix also works for me. This came out of nowhere (seemingly, maybe I overlooked an update to salt-cloud somewhere, but it was working fine for months and started failing yesterday?).

I'm running Ubuntu 14.04 on AWS, and added `master_type: standard` to my `/etc/salt/cloud.providers.d/ec2.conf` under the `minion:` key. 

Thanks. Interested in theories on why this might have broken in the first place if anyone has them.

**Comment by Ligthert** [comment]
I resolved the issue by doing this:
> salt \* file.replace /etc/salt/minion pattern='master_type: str' repl='master_type: standard'

**Comment by dannykansas** [comment]
@Ligthert how are you able to connect to your minions if the salt-minion daemon is unable to start due to that misplaced `str`?

**Comment by rallytime** [comment]
The correct fix for the time being is the one posted above by @complistic-gaff:
```
my-provider-config:
  minion:
    master_type: standard
```
Apologies for the delay in releasing a fix for this, but the bug has been fixed in the upcoming 2015.5.4 release. We are currently running through our QA processes on the release, so hopefully it will be out very soon.

**Comment by Ligthert** [comment]
@dannykansas Before updating my legion of minions I did: `salt \* file.replace /etc/salt/minion pattern='master_type: str' repl='master_type: standard'`
